src/include/miopen/find_solution.hpp

```diff
@@ -163,16 +171,24 @@ template <class Solver, class Context, class Problem, class Db>
 ConvSolution FindSolution(Solver s,
                           const Context& context,
                           const Problem& problem,
-                          Db& db,
+                          Db&& db,
                           const AnyInvokeParams& invoke_ctx,
                           const std::string& perf_cfg               = "",
                           const std::optional<FindOptions>& options = std::nullopt)
 {
     static_assert(sizeof(Solver) == sizeof(SolverBase), "Solver must be stateless");
     static_assert(std::is_base_of<SolverBase, Solver>{}, "Not derived class of SolverBase");
+
+    decltype(auto) db_getter = [&]() -> decltype(auto) {
+        if constexpr(std::is_invocable_v<Db>)
+            return db;
+        else
+            return [&]() -> std::decay_t<Db>& { return db; };
+    }();
```

Comment:
@DrizztDoUrden  Can you please help me understand what is going on here and why ?

Reply:
Yes, sure! This makes a compile-time check of where `Db` is a type with an `operator()` with no parameters. If so, it returns the value in db. Otherwise it returns a lamda that, in turns, returns reference to the value of db. This does not have a check that Db is an actual data base type, but that is implied. I have considered making a pseudo-concept to do the check, but I do not think it is feasible here, the build would fail anyway and code would not be trivial for a proper check as dbs do not have unified API. I can make it if needed to make this lamda architecturally correct.